ribs/emitters/rankers.py

```diff
@@ -0,0 +1,368 @@
+    that resets the internal state of the ranker
+    (e.g. in :class:`ribs.emitters.rankers._random_direction_ranker`).
+
+    Child classes are only required to override :meth:`rank`.
+    """
+
+    @abstractmethod
+    def rank(self, emitter, archive, solution_batch, objective_batch,
+             measure_batch, metadata, add_statuses, add_values):
+        # pylint: disable=missing-function-docstring
+        pass
+
+    # Generates the docstring for rank
+    rank.__doc__ = f"""
+Generates a batch of indicies that represents an ordering of ``solution_batch``.
+
+Args:
+{_args.emitter}
+{_args.archive}
+{_args.solution_batch}
+{_args.objective_batch}
+{_args.measure_batch}
+{_args.metadata}
+{_args.add_statuses}
+{_args.add_values}
+
+Returns:
+{_returns.indices}
+    """
+
+    def reset(self, emitter, archive):
+        # pylint: disable=missing-function-docstring
+        pass
+
+    # Generates the docstring for reset
+    reset.__doc__ = f"""
+Resets the internal state of the ranker.
+
+Args:
+{_args.emitter}
+{_args.archive}
+   """
+
+
+class ImprovementRanker(RankerBase):
+    # TODO implement
+    pass
+
+
+class TwoStageImprovementRanker(RankerBase):
+    """Ranks the solutions based on the improvement in the objective.
+
+    This ranker originates in `Fontaine 2020
+    <https://arxiv.org/abs/1912.02400>`_ in which it was referred to as
+    "Improvement Emitter".
+    This ranker ranks solutions that improves the archive more higher.
+    Moreover, new solutions are ranked above improved ones, which ranks above
+    ones that were not added to the archive.
+    """
+
+    def rank(self, emitter, archive, solution_batch, objective_batch,
+             measure_batch, metadata, add_statuses, add_values):
+        # New solutions sort ahead of improved ones, which sort ahead of ones
+        # that were not added. Note that lexsort sorts the values in ascending
+        # order, so we use numpy fancy indexing to reverse the sorted array.
+        return np.lexsort((add_values, add_statuses))[::-1]
+
+    # Generates the docstring for rank
+    rank.__doc__ = f"""
+Generates a list of indices that represents an ordering of solutions.
+
+Args:
+{_args.emitter}
+{_args.archive}
+{_args.solution_batch}
+{_args.objective_batch}
+{_args.measure_batch}
+{_args.metadata}
+{_args.add_statuses}
+{_args.add_values}
+
+Returns:
+{_returns.indices}
+    """
+
+
+class RandomDirectionRanker(RankerBase):
+    """Ranks the solutions based on projection onto a direction in
+    measure space.
+
+    This ranker originates in `Fontaine 2020
+    <https://arxiv.org/abs/1912.02400>`_ as RandomDirectionEmitter.
+    The solutions are ranked solely based on their projection onto a random
+    direction in measure space.
+
+    To rank the solutions first by whether they were added, and then by
+    the projection, refer to
+    :class:`ribs.emitters.rankers.TwoStageRandomDirectionRanker`.
+    """
+
+    def __init__(self, seed=None):
```

Comment:
I suppose we'd need to make `seed` be part of the `RankerBase.__init__` so that we have a consistent API